--- a/libsodium/src/libsodium/crypto_generichash/blake2/ref/blake2.h
+++ b/libsodium/src/libsodium/crypto_generichash/blake2/ref/blake2.h
@@ -74,7 +74,7 @@
     uint8_t  personal[BLAKE2S_PERSONALBYTES];  // 32
   } blake2s_param;
 
-CRYPTO_ALIGN( 64 ) typedef struct blake2s_state_
+typedef struct CRYPTO_ALIGN( 64 ) blake2s_state_
   {
     uint32_t h[8];
     uint32_t t[2];

--- a/libsodium/src/libsodium/include/sodium/crypto_generichash_blake2b.h
+++ b/libsodium/src/libsodium/include/sodium/crypto_generichash_blake2b.h
@@ -20,7 +20,7 @@
 # pragma pack(push, 1)
 #endif
 
-typedef CRYPTO_ALIGN(64) struct crypto_generichash_blake2b_state {
+typedef struct CRYPTO_ALIGN(64) crypto_generichash_blake2b_state {
     uint64_t h[8];
     uint64_t t[2];
     uint64_t f[2];
