--- a/feeds/luci/modules/luci-base/root/usr/libexec/rpcd/luci
+++ b/feeds/luci/modules/luci-base/root/usr/libexec/rpcd/luci
@@ -46,7 +46,17 @@
 	getCPUInfo = {
 		call = function()
 			local sys = require "luci.sys"
-			return { result = sys.exec("top -n1 | awk '/^CPU/ {printf(\"%d%\",100-$8)}'") }
+			local rv = {}
+			rv.cpufreq = sys.exec("echo A72:$(expr $(cat /sys/devices/system/cpu/cpufreq/policy4/cpuinfo_cur_freq) / 1000) MHz A53:$(expr $(cat /sys/devices/system/cpu/cpufreq/policy0/cpuinfo_cur_freq) / 1000) MHz +$(awk '{printf(\"%.1f°C\",$0/1000)}' /sys/class/thermal/thermal_zone0/temp)")
+			rv.cpufree = sys.exec("top -n1 | awk '/^CPU/ {printf(\"%d%\",100-$8)}'")
+			return rv
+		end
+	},
+
+	getETHInfo = {
+		call = function()
+			local sys = require "luci.sys"
+			return { result = sys.exec("ethinfo") }
 		end
 	},
 

--- a/feeds/luci/modules/luci-mod-status/root/usr/share/rpcd/acl.d/luci-mod-status-index.json
+++ b/feeds/luci/modules/luci-mod-status/root/usr/share/rpcd/acl.d/luci-mod-status-index.json
@@ -19,6 +19,15 @@
 		}
 	},
 
+	"luci-mod-status-index-ethinfo": {
+		"description": "Grant access to ETHInfo status display",
+		"read": {
+			"ubus": {
+				"luci": [ "getETHInfo" ]
+			}
+		}
+	},
+
 	"luci-mod-status-index-dhcp": {
 		"description": "Grant access to DHCP status display",
 		"read": {

--- a/feeds/luci/modules/luci-mod-status/htdocs/luci-static/resources/view/status/include/10_system.js
+++ b/feeds/luci/modules/luci-mod-status/htdocs/luci-static/resources/view/status/include/10_system.js
@@ -61,6 +61,7 @@
 			_('Hostname'),         boardinfo.hostname,
 			_('Model'),            boardinfo.model,
 			_('Architecture'),     boardinfo.system,
+			_('CPU Info'),         cpuinfo.cpufreq,
 			_('Firmware Version'), (L.isObject(boardinfo.release) ? boardinfo.release.description + ' / ' : '') + (luciversion || ''),
 			_('Kernel Version'),   boardinfo.kernel,
 			_('Local Time'),       datestr,
@@ -70,7 +71,7 @@
 				systeminfo.load[1] / 65535.0,
 				systeminfo.load[2] / 65535.0
 			) : null,
-			_('CPU usage (%)'),    cpuinfo.result
+			_('CPU usage (%)'),    cpuinfo.cpufree
 		];
 
 		var table = E('table', { 'class': 'table' });

--- a/package/ipk/luci-app-cpufreq/root/etc/config/cpufreq
+++ b/package/ipk/luci-app-cpufreq/root/etc/config/cpufreq
@@ -1,2 +1,9 @@
 
 config settings 'cpufreq'
+	option mode0 'schedutil'
+	option min0 '816000'
+	option max0 '1608000'
+	option mode4 'schedutil'
+	option min4 '816000'
+	option max4 '2016000'
+
